File: jbake-core/src/test/java/org/jbake/template/ModelExtractorsDocumentTypeListenerTest.java
package org.jbake.template;

import org.jbake.engine.ModelExtractors;
import org.jbake.model.DocumentTypes;
import org.jbake.template.model.ModelExtractorsDocumentTypeListener;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;

public class ModelExtractorsDocumentTypeListenerTest {

    @Test
    public void shouldRegisterExtractorsForCustomType() {
        // given: "A document type is known"
        String newDocumentType = "project";
        DocumentTypes.addDocumentType(newDocumentType);
        ModelExtractorsDocumentTypeListener listener = new ModelExtractorsDocumentTypeListener();

        // when: "the listener is called with that type"
        listener.added(newDocumentType);

        // then: "an extractor is registered by pluralized type as key"
        assertThat(ModelExtractors.getInstance().supportsExtension("projects")).isTrue();

        // and: "an extractor for published types is registered"
        assertThat(ModelExtractors.getInstance().supportsExtension("published_projects")).isTrue();
    }
}

File: jbake-core/src/test/kotlin/org/jbake/template/ModelExtractorsDocumentTypeListenerTest.kt
package org.jbake.template

import org.jbake.model.DocumentTypeRegistry.addDocumentType
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.shouldBe
import org.jbake.model.ModelExtractorsDocumentTypeListener

class ModelExtractorsDocumentTypeListenerTest : StringSpec({
    "shouldRegisterExtractorsForCustomType" {
        // Given: "A document type is known."
        val newDocumentType = "project"
        addDocumentType(newDocumentType)
        val listener = ModelExtractorsDocumentTypeListener()

        // When: "the listener is called with that type."
        listener.onAdded(newDocumentType)

        // Then: "an extractor is registered by pluralized type as key."
        ModelExtractorsRegistry.instance.containsKey("projects") shouldBe true

        // And: "an extractor for published types is registered."
        ModelExtractorsRegistry.instance.containsKey("published_projects") shouldBe true
    }
})
